src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2634,34 +2628,36 @@ void ShenandoahHeap::verify_rem_set_at_mark() {
           // else, object's start is marked dirty and obj is not an objArray, so any interesting pointers are covered
         }
         // else, this humongous object is not marked so no need to verify its internal pointers
-        if (!scanner->verify_registration(obj_addr, obj->size())) {
+        if (!scanner->verify_registration(obj_addr, ctx)) {
           ShenandoahAsserts::print_failure(ShenandoahAsserts::_safe_all, obj, obj_addr, NULL,
                                           "Verify init-mark remembered set violation", "object not properly registered", __FILE__, __LINE__);
         }
       } else if (!r->is_humongous()) {
         HeapWord* t = r->top();
         while (obj_addr < t) {
           oop obj = oop(obj_addr);
-          // ctx->is_marked() returns true if mark bit set (TAMS not relevant here)
+          // ctx->is_marked() returns true if mark bit set or if obj above TAMS.
           if (!ctx || ctx->is_marked(obj)) {
             // For regular objects (not object arrays), if the card holding the start of the object is dirty,
             // we do not need to verify that cards spanning interesting pointers within this object are dirty.
             if (!scanner->is_card_dirty(obj_addr) || obj->is_objArray()) {
               obj->oop_iterate(&check_interesting_pointers);
             }
             // else, object's start is marked dirty and obj is not an objArray, so any interesting pointers are covered
-            if (!scanner->verify_registration(obj_addr, obj->size())) {
+            if (!scanner->verify_registration(obj_addr, ctx)) {
               ShenandoahAsserts::print_failure(ShenandoahAsserts::_safe_all, obj, obj_addr, NULL,
                                             "Verify init-mark remembered set violation", "object not properly registered", __FILE__, __LINE__);
             }
             obj_addr += obj->size();
-          } // Else, this object is not live so we don't verify dirty cards contained therein.
-
-          if (ctx) {
-            // TAMS not relevant here
-            obj_addr = ctx->get_next_marked_addr(obj_addr, t);
           } else {
-            obj_addr += obj->size();
+            // This object is not live so we don't verify dirty cards contained therein
+            ShenandoahHeapRegion* r = heap_region_containing(obj_addr);
+            HeapWord* tams = ctx->top_at_mark_start(r);
+            if (obj_addr >= tams) {
+              obj_addr += obj->size();
```

Comment:
If we verify remset while old marking is "ongoing", ctx is not "reliable".  In that case, I had been arranging for ctx to equal NULL with the understanding that obj->size() would get me to the next (possibly coalesced and filled) object within old-gen.  I don't yet follow the logic for how you can always rely on ctx being not NULL here.

Reply:
If `ctx` is NULL, we'll have gone into the other path and used `obj->size()` to advance `obj_addr`. I'll add an assert that `obj_addr` is below TAMS during the init-mark verification.  We only verify the remembered set during a mixed evacuation or global collect (as of this change). In both cases there should be no concurrent marking of old (I'll add an assert for this also).

Reply:
I was able to simplify this (and similar code in a few other places) because we've previously checked if the object is marked (which would return true for an object above TAMS).